File: src/simulations/Fractal.java
package simulations;

import java.awt.Color;
import java.awt.Graphics;
import java.util.ArrayList;
import java.util.List;

import spacesimulation.*;
import spacesimulation.algebra.*;

public class Fractal extends Simulation {
    CartesianCoordinateSystem cart = new CartesianCoordinateSystem(false, 100, 10, Color.white);
    private Point3d[] corners;
    private List<Point3d> points;

    public Fractal(int dim, Simulator sim) {
        super(sim);
        corners = new Point3d[dim];
        reset();
    }

    @Override
    public void tick() {
        int speed = 5;
        for (int i = 0; i < speed; i++) {
            Point3d corner = corners[(int)(Math.random() * corners.length)];
            points.add(new Point3d(points.get(points.size() - 1).getConnectingVectorTo(corner).scale(0.5).add(points.get(points.size() - 1).getPositionVector())));
        }
        drawer.setCameraAngleHorizontal(0);
        drawer.setCameraAngleVertical(0);
    }

    @Override
    public void render(Graphics g) {
        // cart.render(drawer, g);
        for (Point3d point : corners) drawer.drawDot(point, 4, Color.green, g);
        for (Point3d point : points) drawer.drawDot(point, 1, Color.white, g);
    }

    @Override
    public void reset() {
        double scale = 1000;
        points = new ArrayList<Point3d>();
        // for (int i = 0; i < (int)((double)corners.length / 2.0 + 0.6); i++) corners[i] = new Point3d(Math.random() * scale * 2 - scale, Math.random() * scale * 2 - scale, 0);
        // for (int i = (int)((double)corners.length / 2.0 + 0.6); i < corners.length; i++) corners[i] = new Point3d(corners[i - corners.length / 2].getPositionVector().scale(-1));
        corners[0] = new Point3d(0,-100,0);
        corners[1] = new Point3d(-95,-31,0);
        corners[2] = new Point3d(-59,81,0);
        corners[3] = new Point3d(59,81,0);
        corners[4] = new Point3d(95,-31,0);
        points.add(new Point3d(Math.random() * scale * 2 - scale, Math.random() * scale * 2 - scale, 0));
        drawer.setZoom(4.7);
    }
}

File: src/main/java/simulations/Fractal.kt
package simulations

import framework.Simulation
import algebra.Vec3
import framework.WatchBoolean
import framework.WatchDouble
import framework.WatchInt
import framework.physics.Seconds
import toVec
import kotlin.math.PI
import kotlin.math.cos
import kotlin.math.sin

class Fractal : Simulation("Fractal") {
    @WatchInt("Corners", 3, 10)
    private var dim: Int = 3
        set(value) {
            field = value
            reset()
        }
    private val corners: ArrayList<Vec3> = ArrayList()
    private val points: MutableList<Vec3> = ArrayList()
    @WatchDouble("PointsPerSecond", 1.0, 10000.0)
    private var pointsPerSecond: Double = 100.0
    @WatchBoolean("Skip to End")
    private var skip = false
    private val secondsPerPoint: Double
        get() = 1.0 / pointsPerSecond

    private var secondsSinceLastPoint = 0.0

    init {
        camera.x = 0.0
        camera.y = 0.0
        camera.z = 10.0
        camera.focalLength = 1.0
        camera.zoom = 0.022
        camera.phi = 0.0
        camera.theta = PI
        reset()
    }

    override fun tick(dt: Seconds) {
        if (!isInitialized()) return
        if (!skip) {
            secondsSinceLastPoint += dt
            if (secondsSinceLastPoint < secondsPerPoint) return

            secondsSinceLastPoint = 0.0
        }
        val corner = corners.random()
        val last = points.last()
        val newPoint = last + (corner - last) / 2
        synchronized(points) {
            points.add(newPoint)
        }
    }

    private fun isInitialized(): Boolean = corners.isNotEmpty()

    override fun render() {
        val color = Conf.colorScheme.smallObjectColor.toVec()
        synchronized(points) {
            points.indices.forEach { camera.renderPixel(points[it], color, null) }
        }
    }

    override fun reset() {
        val radius = 100.0
        synchronized(points) {
            points.clear()
            corners.clear()
            for (i in 0 until dim) {
                val phi = 2 * Math.PI / dim * i
                val x = radius * cos(phi + Math.PI / 2)
                val y = radius * sin(phi + Math.PI / 2)
                corners.add(Vec3(x, y, 0.0))
            }
            points.add(Vec3(Math.random() * radius * 2 - radius, Math.random() * radius * 2 - radius, 0.0))
        }
    }
}

fun main() {
    Fractal().start()
}